### PR TITLE
OHBufferedMutator in OBKV Hbase 1_x_comp

### PR DESCRIPTION
**Pull request by JackShi148**
OHBufferedMutator in OBKV Hbase 1_x_comp

## Summary

Add OHBufferedMutator in OBKV Hbase 1_x_comp. The usage is the same as 1.3.6 hbase, which is asynchronous thread-safe batch operation. 
Limitations: 
- 1. Only support Delete and Put.
- 2. Only support the same family in one batch.
- 3. Only support the same type operations in one batch.

### CODE REVIEW

File: src/test/java/com/alipay/oceanbase/hbase/OHBufferedMutatorTest.java [hunk 1]
```diff
@@ -0,0 +1,471 @@
+package com.alipay.oceanbase.hbase;
+
```
Comment (stuBirdFly) [inline src/test/java/com/alipay/oceanbase/hbase/OHBufferedMutatorTest.java:2]: 由于这个是connection的东西，考虑放在OHConnectionTest里？

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 2]
```diff
@@ -17,53 +17,385 @@
...
+                // in concurrent situation, asyncWriteBuffer may be empty here
+                // for other threads flush all buffer
+                if (execBuffer.isEmpty()) {
+                    break;
+                }
+                // namespace n1, n1:table_name
+                // namespace default, table_name
+                String tableNameString = tableName.getNameAsString();
+                // for now, operations' family is the same
+                byte[] family = execBuffer.getFirst().getFamilyMap().firstKey();
```
Comment (stuBirdFly) [inline src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:234]: 这些东西有没有考虑空指针的问题

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 3]
```diff
@@ -17,53 +17,385 @@
...
+                ObTableBatchOperation batch = buildObTableBatchOperation(execBuffer);
+                // table_name$cf_name
+                String targetTableName = getTargetTableName(tableNameString, Bytes.toString(family));
+                ObTableBatchOperationRequest request = buildObTableBatchOperationRequest(batch, targetTableName, pool);
+                ObTableBatchOperationResult result = (ObTableBatchOperationResult) obTableClient.execute(request);
+            }
+        } catch (Exception ex) {
+            LOGGER.error(LCD.convert("01-00026"), ex);
```
Comment (stuBirdFly) [inline src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:242]: 如果是有些操作失败了，是不是考虑打debug级别的日志？

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 4]
```diff
@@ -17,53 +17,385 @@
...
+            // TODO: need to collect error information and actions during batch operations
+            // TODO: maybe keep in ObTableBatchOperationResult
+            List<Throwable> throwables = new ArrayList<Throwable>();
+            List<Row> actions = new ArrayList<Row>();
+            List<String> addresses = new ArrayList<String>();
+            throwables.add(ex);
+            RetriesExhaustedWithDetailsException error = new RetriesExhaustedWithDetailsException(
```
Comment (stuBirdFly) [inline src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:249]: 如果是操作失败了，需不需要将其放回去，否则操作丢失？

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 5]
```diff
@@ -17,53 +17,385 @@
...
     @Override
     public void close() throws IOException {
-
+        if (closed) {
+            return;
+        }
+        try {
+            asyncExecute(true);
+        } finally {
+            this.pool.shutdown();
```
Comment (stuBirdFly) [inline src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:268]: 这里和hbase保持一致吧，如果flush失败了，则不走到这一步，不放在finally里

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 6]
```diff
@@ -17,53 +17,385 @@
...
+            closed = true;
+        }
     }
 
+    /**
+     * Force to commit all operations
+     */
     @Override
     public void flush() throws IOException {
-
```
Comment (stuBirdFly) [inline src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:62]: flush手动触发，将剩余的提交上去，和hbase一样，不检查是否已经关闭

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 7]
```diff
@@ -17,53 +17,385 @@
...
+        if (closed) {
```
Comment (stuBirdFly) [inline src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:287]: flush手动触发，将剩余的提交上去，和hbase一样，不检查是否已经关闭

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 8]
```diff
@@ -17,53 +17,385 @@
...
+        for (Mutation mutation : execBuffer) {
+            checkFamilyViolation(mutation.getFamilyMap().keySet());
+            for (Map.Entry<byte[], List<KeyValue>> entry : mutation.getFamilyMap().entrySet()) {
+                keyValueList.addAll(entry.getValue());
+            }
+        }
+        return buildObTableBatchOperation(keyValueList, false, null);
+    }
+
+    private ObTableBatchOperation buildObTableBatchOperation(List<KeyValue> keyValueList,
```
Comment (stuBirdFly) [inline src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:309]: 考虑调用ohtable的接口，没必要重复写。可以将table的写为static public

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 9]
```diff
@@ -17,53 +17,385 @@
...
+                qualifiers.add(kv.getQualifier());
+            }
+            batch.addTableOperation(buildObTableOperation(kv, putToAppend));
+        }
+        batch.setSameType(true);
+        batch.setSamePropertiesNames(true);
+        return batch;
+    }
+
+    private ObTableOperation buildObTableOperation(KeyValue kv, boolean putToAppend) {
```
Comment (stuBirdFly) [inline src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:324]: 同上？

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 10]
```diff
@@ -17,53 +17,385 @@
...
+                    new Object[] { kv.getRow(), kv.getQualifier(), -kv.getTimestamp() }, null, null);
+            case DeleteFamily:
+                return getInstance(DEL, new Object[] { kv.getRow(), null, -kv.getTimestamp() },
+                    null, null);
+            default:
+                throw new IllegalArgumentException("illegal mutation type " + kvType);
+        }
+    }
+
+    private ObTableBatchOperationRequest buildObTableBatchOperationRequest(ObTableBatchOperation obTableBatchOperation,
```
Comment (stuBirdFly) [inline src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:351]: 同上

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 11]
```diff
@@ -17,53 +17,385 @@
...
+        ObTableBatchOperationRequest request = new ObTableBatchOperationRequest();
+        request.setTableName(targetTableName);
+        request.setReturningAffectedRows(true);
+        request.setEntityType(ObTableEntityType.HKV);
+        request.setBatchOperation(obTableBatchOperation);
+        request.setPool(pool);
+        return request;
+    }
+
+    private void checkFamilyViolation(Collection<byte[]> families) {
```
Comment (stuBirdFly) [inline src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:363]: 同上，如果多的话，可以将这些封装成一个工具类

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 12]
```diff
@@ -17,53 +17,385 @@
...
+                throw new IllegalArgumentException("family is empty");
+            }
+            if (isBlank(Bytes.toString(family))) {
+                throw new IllegalArgumentException("family is blank");
+            }
+        }
+
+    }
+
+    private String getTargetTableName(String tableNameString, String familyString) {
```
Comment (stuBirdFly) [inline src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:383]: 同上

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 13]
```diff
@@ -17,53 +17,385 @@
...
+        checkArgument(tableNameString != null, "tableNameString is null");
+        checkArgument(familyString != null, "familyString is null");
+        if (conf.getBoolean(HBASE_HTABLE_TEST_LOAD_ENABLE, false)) {
+            return getTestLoadTargetTableName(tableNameString, familyString);
+        }
+        return getNormalTargetTableName(tableNameString, familyString);
+    }
+
+    private String getTestLoadTargetTableName(String tableNameString, String familyString) {
```
Comment (stuBirdFly) [inline src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:392]: 同上

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 14]
```diff
@@ -17,53 +17,385 @@
...
+        String suffix = conf.get(HBASE_HTABLE_TEST_LOAD_SUFFIX,
+            DEFAULT_HBASE_HTABLE_TEST_LOAD_SUFFIX);
+        return tableNameString + suffix + "$" + familyString;
+    }
+
+    private String getNormalTargetTableName(String tableNameString, String familyString) {
```
Comment (stuBirdFly) [inline src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:398]: 同上

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java (hunk 4 above)
Comment (JackShi148) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:249]: 无法追溯是哪个报错

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 15]
```diff
@@ -17,53 +17,385 @@
...
+    private static final int                BUFFERED_PARAM_UNSET   = -1;
+
+    private final ExceptionListener         listener;
+
+    protected final ObTableClient           obTableClient;
+    private final TableName                 tableName;
+    private volatile Configuration          conf;
+    private final OHConnectionConfiguration connectionConfig;
+
+    @VisibleForTesting
```
[the remark this answers is not in the document]
Comment (JackShi148) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java]: done

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 16]
```diff
@@ -17,53 +17,385 @@
...
+import static com.alipay.oceanbase.rpc.util.TableClientLoggerFactory.LCD;
+import static org.apache.commons.lang.StringUtils.isBlank;
 
 @InterfaceAudience.Private
 public class OHBufferedMutatorImpl implements BufferedMutator {
-    public OHBufferedMutatorImpl(OHConnectionImpl ohConnection, BufferedMutatorParams params) {
+    private static final Logger             LOGGER                 = TableHBaseLoggerFactory
+                                                                       .getLogger(OHBufferedMutatorImpl.class);
+
+    private static final int                BUFFERED_PARAM_UNSET   = -1;
```
[the remark this answers is not in the document]
Comment (JackShi148) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java]: Set BUFFERED_PARAM_UNSET public in OHConnectionImpl.

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 17]
```diff
@@ -17,53 +17,385 @@
...
+        long toAddSize = 0;
+        // check if every mutation's family is the same
+        if (!validateSameFamily(mutations)) {
+            throw new IllegalArgumentException("Family should keep the same in one batch.");
+        }
+        for (Mutation m : mutations) {
+            if (!validateInsUpAndDelete(m)) {
+                throw new IllegalArgumentException("Only support for Put and Delete for now.");
+            }
+            if (m instanceof Put) {
```
[the remark this answers is not in the document]
Comment (JackShi148) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java]: done.

File: src/test/java/com/alipay/oceanbase/hbase/OHBufferedMutatorTest.java (hunk 1 above)
Comment (JackShi148) [reply src/test/java/com/alipay/oceanbase/hbase/OHBufferedMutatorTest.java]: done.

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java (hunk 2 above)
Comment (JackShi148) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java]: Yes, TableName's constructor will ensure that the nameAsString is not empty and the execBuffer's elements have been checked for their family.

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java (hunk 7 above)
Comment (JackShi148) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java]: done.

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java (hunk 8 above)
Comment (JackShi148) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java]: done.

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java (hunk 9 above)
Comment (JackShi148) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java]: done.

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java (hunk 10 above)
Comment (JackShi148) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java]: done.

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java (hunk 14 above)
Comment (JackShi148) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java:333]: Maybe it still needs to use the this.conf? Can remove after removing this.conf.

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 18]
```diff
@@ -17,53 +17,320 @@
...
+
+    final ConcurrentLinkedQueue<Mutation>   asyncWriteBuffer       = new ConcurrentLinkedQueue<Mutation>();
+    AtomicLong                              currentAsyncBufferSize = new AtomicLong(0);
+
+    private final long                      writeBufferSize;
+    private final int                       maxKeyValueSize;
+    private boolean                         closed                 = false;
+    private final ExecutorService           pool;
+    private final int                       rpcTimeout;
+    private final int                       operationTimeout;
```
[the remark this answers is not in the document]
Comment (JackShi148) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java]: done.

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 19]
```diff
@@ -17,53 +17,320 @@
...
+                }
+                try {
+                    ObTableBatchOperationResult result = (ObTableBatchOperationResult) obTableClient.execute(request);
+                } catch (Exception ex) {
+                    LOGGER.debug("Errors occur during mutation operation", ex);
+                    try {
+                        // retry every single operation
+                        while (!execBuffer.isEmpty()) {
+                            // poll elements from execBuffer to recollect remaining operations
+                            Mutation m = execBuffer.poll();
```
[the remark this answers is not in the document]
Comment (JackShi148) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java]: done.

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 20]
```diff
@@ -17,53 +17,320 @@
...
+                    checkAllOpsIsSameType(execBuffer);
+                    // for now, operations' family is the same
+                    byte[] family = execBuffer.getFirst().getFamilyMap().firstKey();
+                    ObTableBatchOperation batch = buildObTableBatchOperation(execBuffer);
+                    // table_name$cf_name
+                    String targetTableName = getTargetTableName(tableNameString, Bytes.toString(family));
+                    request = OHTable.buildObTableBatchOperationRequest(batch, targetTableName, pool);
+                } catch (Exception ex) {
+                    LOGGER.error("Errors occur before mutation operation", ex);
+                    throw ex;
```
[the remark this answers is not in the document]
Comment (JackShi148) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java]: done.

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java [hunk 21]
```diff
@@ -17,53 +17,320 @@
...
+        return OHTable.buildObTableBatchOperation(keyValueList, false, null);
+    }
+
+    private String getTargetTableName(String tableNameString, String familyString) {
+        checkArgument(tableNameString != null, "tableNameString is null");
+        checkArgument(familyString != null, "familyString is null");
+        if (conf.getBoolean(HBASE_HTABLE_TEST_LOAD_ENABLE, false)) {
+            return getTestLoadTargetTableName(tableNameString, familyString);
+        }
+        return getNormalTargetTableName(tableNameString, familyString);
```
[the remark this answers is not in the document]
Comment (JackShi148) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBufferedMutatorImpl.java]: done.

**Comment by shenyunlong** [comment]
LGTM

**Pull request merged**